pkg/intermediate/aggregate.go

```diff
@@ -0,0 +1,284 @@
+	workerPool chan chan *entities.Message
+	// messageChan is the channel to receive the message
+	messageChan chan *entities.Message
+	// workerNum is the number of workers to process the messages
+	workerNum int
+	// workerList is the list of workers
+	workerList []*worker
+	// correlateFields are the fields to be filled in correlating process
+	correlateFields []string
+}
+
+type Tuple struct {
+	SourceAddress      string
+	DestinationAddress string
+	Protocol           uint8
+	SourcePort         uint16
+	DestinationPort    uint16
+}
+
+func InitAggregationProcess(messageChan chan *entities.Message, workerNum int, correlateFields []string) (*aggregation, error) {
+	if messageChan == nil {
+		return nil, fmt.Errorf("Cannot create aggregation process without message channel.")
+	} else if workerNum <= 0 {
+		return nil, fmt.Errorf("Worker number cannot be <= 0.")
+	}
+	return &aggregation{
+		make(map[Tuple][]entities.Record),
+		sync.RWMutex{},
+		make(chan chan *entities.Message),
+		messageChan,
+		workerNum,
+		make([]*worker, 0),
+		correlateFields,
+	}, nil
+}
+
+func (a *aggregation) Start() {
+	for i := 0; i < a.workerNum; i++ {
+		w := createWorker(i, a.workerPool, a.AggregateMsgBy5Tuple)
+		w.start()
+		a.workerList = append(a.workerList, w)
+	}
+	for message := range a.messageChan {
+		channel := <-a.workerPool
+		channel <- message
+	}
+}
+
+func (a *aggregation) Stop() {
+	for _, worker := range a.workerList {
+		worker.stop()
+	}
+}
+
+// AggregateMsgBy5Tuple gets 5-tuple info from records in message and stores in cache
+func (a *aggregation) AggregateMsgBy5Tuple(message *entities.Message) error {
+	addOriginalExporterInfo(message)
+	if message.Set.GetSetType() == entities.Template { // skip template records
+		return nil
+	}
+	records := message.Set.GetRecords()
+	for _, record := range records {
+		tuple, err := getTupleFromRecord(record)
+		if err != nil {
+			return err
+		}
+		a.correlateRecords(tuple, record)
+	}
+	return nil
+}
+
+func (a *aggregation) GetTupleRecordMap() map[Tuple][]entities.Record {
+	a.tupleRecordLock.RLock()
+	defer a.tupleRecordLock.RUnlock()
+	return a.tupleRecordMap
+}
+
+func (a *aggregation) DeleteTupleFromMap(tuple Tuple) {
+	a.tupleRecordLock.Lock()
+	defer a.tupleRecordLock.Unlock()
+	delete(a.tupleRecordMap, tuple)
+}
+
+// correlateRecords fills records info by correlating incoming and current records
+func (a *aggregation) correlateRecords(tuple Tuple, record entities.Record) {
+	existingRecords := a.GetTupleRecordMap()[tuple]
+	// only fill the information for record from source node
+	if isRecordFromSrc(record) {
+		var isFilled bool
+		for _, existingRec := range existingRecords {
+			for _, field := range a.correlateFields {
+				if containsInfoElement(record.GetInfoElementMap(), field) {
+					record.GetInfoElementMap()[field].Value = existingRec.GetInfoElementMap()[field].Value
+					isFilled = true
+				}
+			}
+			if isFilled {
+				break
```

Comment:
Why break here?
Is that possible the incoming record (fromSrc) gets field_A from existingRec_1 (fromDst) and field_B from existingRec_2 (fromDst)?
Or all fromSrc records always have the same set of fields populated? Similar for all fromDst records?

Reply:
Currently we don't have use cases that incoming record(fromSrc) will need fill the fields from different dst records. One fromSrc and one fromDst will complete each other. But it is a good point to pay attention in the future.